### DISCUSSION TITLE
Q: How to store trained model using tf.saved_model.builder.SavedModelBuilder ?

### DISCUSSION THREAD

**Discussion by LukaszCmielowski** [horovod/horovod#2751] [Q&A]
Do you have any example how to use SavedModelBuilder (SavedModel) to store the model after training?.
Most of sample are using checkpoint that is not enough in some cases since it requires knowledge about graph.

**Comment by atinsood** [comment]
@alsrgv do you mind taking a look, we are interested in this as well. It might be just the lack of knowledge of TF in our case. But I am assuming we an hook this up in the monitored session and run this only as chief, something along those lines, but an example would be helpful

**Comment by atinsood** [comment]
```if hvd.rank() == 0:
        classification_inputs = tf.saved_model.utils.build_tensor_info(image)
        classification_outputs_classes = tf.saved_model.utils.build_tensor_info(predict)
        classification_signature = (
            tf.saved_model.signature_def_utils.build_signature_def(
                inputs={tf.saved_model.signature_constants.CLASSIFY_INPUTS: classification_inputs},
                outputs={tf.saved_model.signature_constants.CLASSIFY_OUTPUT_CLASSES: classification_outputs_classes},
                method_name=tf.saved_model.signature_constants.CLASSIFY_METHOD_NAME)
            )
        print("classification_signature content:")
        print(classification_signature)

        latest_checkpoint = tf.train.latest_checkpoint(checkpoint_dir)
        saver = tf.train.Saver()

        with tf.Session() as sess:
            sess.run([tf.local_variables_initializer(), tf.tables_initializer()])
            saver.restore(sess, latest_checkpoint)
            builder = tf.saved_model.builder.SavedModelBuilder(os.path.join(checkpoint_dir, 'export'))
            legacy_init_op = tf.group(tf.tables_initializer(), name='legacy_init_op')
            builder.add_meta_graph_and_variables(sess, [tf.saved_model.tag_constants.SERVING], signature_def_map={'predict_images': classification_signature}, legacy_init_op=legacy_init_op)
            builder.save()```

@alsrgv does this feel right?  basically add this to your tensorflow_mnist.py example. does that seems fair?

**Comment by alsrgv** [comment]
@atinsood, thanks for this example. Let me try it and confirm.

**Comment by alsrgv** [comment]
@atinsood, I have tried your example, but it looks like exported saved_model still has HorovodAllreduce operations, which would be a problem if you try to use this graph for inference.

The current approach for inference based on checkpoints is outlined [here](https://github.com/uber/horovod/blob/master/docs/inference.md), and it uses "optimize for inference" operation, which removes training operations from the graph.

It seems that this "optimize for inference" operation for saved_model is not supported yet, https://stackoverflow.com/questions/46239139/tensorflow-optimize-for-inference-a-savedmodel-exported-by-estimator.

We have the same need as you do internally, so we will look into making a script that achieves that purpose, but contributions are welcome!

**Comment by LukaszCmielowski** [comment]
@alsrgv  Is removal of training ops a must have to be able to load such model back ?
Would it work if our deployment service has horovod installed ?
We serve models in Cloud and we have the possibility to pre-install whatever is needed on our site before serving - this basically goes to proper docker image configuration.

**Comment by alsrgv** [comment]
@LukaszCmielowski, it may be possible to LD_PRELOAD Horovod shared library to make serving aware of new operations. You can try that and we can debug together if it doesn't work right off the bat. Meanwhile, we are still planning to figure out how to properly optimizer saved model for inference and remove all the training ops.

**Comment by dmwendt** [comment]
@alsrgv I'm trying to follow the [optimized-for-inference](https://github.com/uber/horovod/blob/master/docs/inference.md) instructions but cannot figure it out. Would it be possible to paste a snippet which shows how to use this tool based on the [tensorflow_mnist.py](https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py) example?  I tried passing in the checkpoint `model.ckpt-100.meta` file but got this error:
```
[libprotobuf ERROR google/protobuf/wire_format_lite.cc:629] String field 'tensorflow.NodeDef.op' contains invalid UTF-8 data when parsing a protocol buffer. 
Use the 'bytes' type if you intend to send raw bytes.
```

**Comment by alsrgv** [comment]
@dmwendt, I had to fight these scripts a little bit, checkpoint saver makes graph definition in text format and freeze graph takes binary format.  Here's an example:

```
$ wget https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/python/tools/optimize_for_inference.py
$ wget https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/python/tools/freeze_graph.py
# grab pbtxt -> pb converter that I made
$ wget https://gist.githubusercontent.com/alsrgv/1a83c7953cb04676f6a0b3ec3917a18c/raw/11451fb09867f71b9c2dde62ff3a1afedd138fbd/pbtxt_to_pb.py
# execute
$ python optimize_for_inference.py --input checkpoints/graph.pbtxt --output optimized_graph.pbtxt --frozen_graph false --input_names input/image --output_names fully_connected_1/BiasAdd
$ python pbtxt_to_pb.py optimized_graph
Converted optimized_graph.pbtxt to optimized_graph.pb
$ python freeze_graph.py --input_graph optimized_graph.pb --input_checkpoint checkpoints/model.ckpt-2 --output_graph frozen_graph.pb --input_binary True --output_node_names fully_connected_1/BiasAdd
2018-03-14 19:50:34.076580: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.2 AVX AVX2 FMA
Converted 8 variables to const ops.

**Comment by dmwendt** [comment]
Ok, this helps. I added the following lines to the end of the [tensorflow_mnist.py](https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py) example:
```
    if hvd.rank() != 0:
       return
    checkpoint_file = tf.train.latest_checkpoint(checkpoint_dir)
    optGraph = optimize_for_inference_lib.optimize_for_inference(tf.get_default_graph().as_graph_def(),
                         ["input/image"], ["fully_connected_1/BiasAdd"],dtypes.float32.as_datatype_enum)
    frozenGraph = freeze_graph.freeze_graph_with_def_protos(optGraph,None,
                         checkpoint_file,"fully_connected_1/BiasAdd",None,None,"frozen.pb",True,None)
    #
    with tf.Graph().as_default():
       importer.import_graph_def(frozenGraph, name="")
       with tf.Session() as sess:
          image = tf.get_default_graph().get_tensor_by_name("input/image:0")
          model = tf.get_default_graph().get_tensor_by_name("fully_connected_1/BiasAdd:0")
          predictor = tf.argmax(model, 1, name="predictor")
          inputs_classes = tf.saved_model.utils.build_tensor_info(image)       # input an image
          outputs_classes = tf.saved_model.utils.build_tensor_info(predictor)  # output its class (0-9)
          signature = (.....)
          builder = tf.saved_model.builder.SavedModelBuilder(....)
          legacy_init_op = tf.group(tf.tables_initializer(), name='legacy_init_op')
          builder.add_meta_graph_and_variables(sess,
                                               [tf.saved_model.tag_constants.SERVING],
                                               signature_def_map={'predict_images': signature},
                                               legacy_init_op=legacy_init_op)
          builder.save()
```
And then I was able to verify the inference of the saved model using the `saved_model_cli` tool.
So I traded the two scripts with the equivalent calls to `optimize_for_inference_lib.optimize_for_inference` and `freeze_graph.freeze_graph_with_def_protos`.
 @alsrgv Does this look ok to you? What do you think?

**Comment by alsrgv** [comment]
@dmwendt, this looks good.  Can you publish it as a gist and reference here, so other folks can easily follow?  We will also release a similar example.

**Comment by dmwendt** [comment]
Here is a gist that shows the above workaround (optimize-graph) as well as alternate method.
https://gist.github.com/dmwendt/ed2779f07aa849eda2e1756cd3b9fcb0

**Comment by Magurosenbei** [comment]
Is there an example for tf.keras models?

**Comment by nikhila0912** [comment]
Hi, is there an example with tensorflow estimators? and how do we get the features (like image, model, input_classes and output_classes) for any other model, like bert model
